indexing-service/src/main/java/io/druid/indexing/overlord/http/OverlordResource.java

```diff
@@ -828,34 +903,48 @@ public TaskStatusPlus apply(TaskRunnerWorkItem workItem)
     }
   }
 
-  private Collection<? extends TaskRunnerWorkItem> securedTaskRunnerWorkItem(
-      Collection<? extends TaskRunnerWorkItem> collectionToFilter,
+  private List<TaskStatusPlus> securedTaskStatusPlus(
+      List<TaskStatusPlus> collectionToFilter,
+      @Nullable String dataSource,
+      @Nullable String type,
       HttpServletRequest req
   )
   {
-    Function<TaskRunnerWorkItem, Iterable<ResourceAction>> raGenerator = taskRunnerWorkItem -> {
-      final String taskId = taskRunnerWorkItem.getTaskId();
-      final Optional<Task> optionalTask = taskStorageQueryAdapter.getTask(taskId);
-      if (!optionalTask.isPresent()) {
+    Function<TaskStatusPlus, Iterable<ResourceAction>> raGenerator = taskStatusPlus -> {
+      final String taskId = taskStatusPlus.getId();
+      final String taskDatasource = taskStatusPlus.getDataSource();
+      if (taskDatasource == null) {
         throw new WebApplicationException(
             Response.serverError().entity(
                 StringUtils.format("No task information found for task with id: [%s]", taskId)
             ).build()
         );
       }
-
       return Lists.newArrayList(
           new ResourceAction(
-              new Resource(optionalTask.get().getDataSource(), ResourceType.DATASOURCE),
+              new Resource(taskDatasource, ResourceType.DATASOURCE),
               Action.READ
           )
       );
     };
-
+    List<TaskStatusPlus> dataSourceFilter = collectionToFilter;
+    if (dataSource != null) {
```

Comment:
this block can be removed now since the datasource filtering is handled by the metadata store now

you can also have a more optimized authorization check path for the `datasource != null` case now, where you don't need to run the auth check on all the tasks returned (since they'll all have the same datasource), and just check that the user has access to the datasource once

Reply:
makes sense, done.